File: src/main/java/com/zhongan/devpilot/completions/inline/render/InlineElementRenderer.java
package com.zhongan.devpilot.completions.inline.render;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.EditorCustomElementRenderer;
import com.intellij.openapi.editor.Inlay;
import com.intellij.openapi.editor.markup.TextAttributes;
import com.zhongan.devpilot.completions.inline.CompletionPreview;

import java.awt.Color;
import java.awt.Graphics;
import java.awt.Rectangle;

import org.jetbrains.annotations.TestOnly;

public class InlineElementRenderer implements EditorCustomElementRenderer {
    private final Editor editor;

    private final String suffix;

    private boolean deprecated;

    private Color color;

    public InlineElementRenderer(Editor editor, String suffix, boolean deprecated, boolean needHint) {
        this.editor = editor;
        this.suffix = needHint ? suffix + "       " + CompletionPreview.byLineAcceptHintText() : suffix;
        this.deprecated = deprecated;
    }

    public InlineElementRenderer(Editor editor, String suffix, boolean deprecated) {
        this(editor, suffix, deprecated, false);
    }

    @Override
    public int calcWidthInPixels(Inlay inlay) {
        return editor.getContentComponent()
                .getFontMetrics(GraphicsUtils.getFont(editor, suffix)).stringWidth(suffix);
    }

    @TestOnly
    public String getContent() {
        return suffix;
    }

    @Override
    public void paint(Inlay inlay, Graphics g, Rectangle targetRegion, TextAttributes textAttributes) {
        color = color != null ? color : GraphicsUtils.getColor();
        g.setColor(color);
        g.setFont(GraphicsUtils.getFont(editor, suffix));
        g.drawString(suffix, targetRegion.x, targetRegion.y + editor.getAscent());
    }
}

File: src/main/java/com/zhongan/devpilot/completions/inline/render/InlineElementRenderer.kt
package com.zhongan.devpilot.completions.inline.render

import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.EditorCustomElementRenderer
import com.intellij.openapi.editor.Inlay
import com.intellij.openapi.editor.markup.TextAttributes
import org.jetbrains.annotations.TestOnly
import java.awt.Color
import java.awt.Graphics
import java.awt.Rectangle

class InlineElementRenderer(private val editor: Editor, private val suffix: String, private val deprecated: Boolean) :
        EditorCustomElementRenderer {
    private var color: Color? = null
    override fun calcWidthInPixels(inlay: Inlay<*>): Int {
        return editor.contentComponent
                .getFontMetrics(GraphicsUtils.getFont(editor, deprecated)).stringWidth(suffix)
    }

    @TestOnly
    fun getContent(): String {
        return suffix
    }

    override fun paint(
            inlay: Inlay<*>,
            g: Graphics,
            targetRegion: Rectangle,
            textAttributes: TextAttributes
    ) {
        color = color ?: GraphicsUtils.color
        g.color = color
        g.font = GraphicsUtils.getFont(editor, deprecated)
        g.drawString(suffix, targetRegion.x, targetRegion.y + editor.ascent)
    }
}
